src/python/WMCore/WorkQueue/WorkQueue.py

```diff
@@ -818,95 +858,60 @@ def pullWork(self, resources=None):
-                        raise RuntimeError("WMSpec doesn't define policyName, current value: '%s'" % policyName)
-
-                    policyInstance = startPolicy(policyName, self.params['SplittingMapping'],
-                                                 rucioObj=rucioObj, logger=self.logger)
-                    if not policyInstance.supportsWorkAddition():
-                        continue
-                    if policyInstance.newDataAvailable(topLevelTask, element):
-                        skipElement = True
-                        self.backend.updateInboxElements(element.id, TimestampFoundNewData=currentTime)
-                        msg = "There are blocks still open for writing in DBS."
-                        self.logdb.post(element['RequestName'], msg, "warning")
-                        break
-                if skipElement:
-                    continue
+        workflowsToCheck = self.backend.getInboxElements(OpenForNewData=True)
+        self.logger.info("Retrieved a list of %d open workflows", len(workflowsToCheck))
+        workflowsToClose = []
+        currentTime = time.time()
+        for element in workflowsToCheck:
+            # Easy check, close elements with no defined OpenRunningTimeout
+            policy = element.get('StartPolicy', {})
+            openRunningTimeout = policy.get('OpenRunningTimeout', 0)
+            if not openRunningTimeout:
+                self.logger.info("Workflow %s has no OpenRunningTimeout. Queuing to be closed.",
+                                 element['RequestName'])
+                workflowsToClose.append(element.id)
+                continue
 
-                # Check if the delay has passed
-                newDataFoundTime = element.get('TimestampFoundNewData', 0)
-                childrenElements = self.backend.getElementsForParent(element)
-                if childrenElements:
-                    lastUpdate = float(max(childrenElements, key=lambda x: x.timestamp).timestamp)
-                    if (currentTime - max(newDataFoundTime, lastUpdate)) > openRunningTimeout:
-                        workflowsToClose.append(element.id)
-                    # if it is successful remove previous error
-                    self.logdb.delete(element.id, "error", this_thread=True)
-                else:
-                    msg = "ChildElement is empty for element id %s: investigate" % element.id
-                    self.logdb.post(element.id, msg, "error")
-                    # self.logdb.upload2central(element.id)
-                    self.logger.error(msg)
+            # Check if the delay has passed
+            foundNewDataTime = element.get('TimestampFoundNewData', 0)
+            if (currentTime - foundNewDataTime) > openRunningTimeout:
+                self.logger.info("Workflow %s has expired OpenRunningTimeout. Queuing to be closed.",
+                                 element['RequestName'])
+                workflowsToClose.append(element.id)
```

Comment:
should you check that given `element.id` is not already in `workflowsToClose`?

Reply:
Each workflow has only one inbox element, so it's pretty safe. Unless we make a mistake and consider the same workflow multiple times, as I have in the current code. Updating it. Thanks for spotting it.